File: validate-binary-search-tree.cpp
#include <limits>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
	bool isValidBST(TreeNode *root) {
		return isValid(root, INT_MIN, INT_MAX);
	}

private:
	bool isValid(TreeNode *node) {
		if (node == nullptr) return true;
		if (node->val > right || left > node->val) return false;

		return isValid(node->left, left, node->val) && isValid(node->right, node->val, right);
	}	
};

File: validate-binary-search-tree.py
from typing import Optional, Self

class TreeNode:
    def __init__(self, val: int = 0, left: Optional[Self] = None, right: Optional[Self] = None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        
        def isValid(node: TreeNode, left: int, right: int) -> bool:
            if node is None: return True
            if node.val > right or left > node.val: return False

            return valid(node.left, left, node.val) and valid(node.right, node.val, right)

        return isValid(root, -float("inf"), float("inf"))
